### PR TITLE
mpl: Fix error handling in mpl related to shm

### PR DESCRIPTION
**Pull request by jain-surabhi-23**
Check the return value of POSIX calls like `mmap, open`.
Make sure MPL doesnt return `mpi_errno` and `mpi_errno` is set correctly after MPL_shm calls.

### CODE REVIEW

**Comment by jain-surabhi-23** [comment]
@minsii Can you please review this PR?

**Comment by jain-surabhi-23** [comment]
test:intel:all/ofi/all/all/all/all/debug,default

**Comment by minsii** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/all

**Comment by minsii** [comment]
My commit yesterday breaks the ch4 builds. Could you please rebase this PR against #3263, which fixes the issue ?

**Comment by minsii** [comment]
ch3 fails seems related to collective changes:
```
src/mpi/coll/iallreduce/iallreduce_intra_tree_kary.c: In function ‘MPIR_Iallreduce_intra_tree_kary’:
src/mpi/coll/iallreduce/iallreduce_intra_tree_kary.c:30:71: error: ‘TREE_TYPE_KARY’ undeclared (first use in this function)
src/mpi/coll/iallreduce/iallreduce_intra_tree_kary.c:30:71: note: each undeclared identifier is reported only once for each function it appears in
make[2]: *** [src/mpi/coll/iallreduce/lib_libmpi_la-iallreduce_intra_tree_kary.lo] Error 1
```

**Comment by jain-surabhi-23** [comment]
Actually I will update this PR, when #3263 is merged. Is that Ok?
The ch3 failures in collective tests seems weird to me, as it should be unrelated.

**Comment by raffenet** [comment]
Those collective compile failures should be fixed.

**Comment by jain-surabhi-23** [comment]
Yes. I will look into those.

File: src/mpl/src/shm/mpl_shm_mmap.c [hunk 1]
```diff
@@ -133,7 +134,7 @@ int MPL_shm_seg_open(MPL_shm_hnd_t hnd, intptr_t seg_sz)
 int MPL_shm_seg_create_and_attach(MPL_shm_hnd_t hnd, intptr_t seg_sz,
                                   char **shm_addr_ptr, int offset)
 {
-    int rc = 0;
+    int rc = -1;
```
Comment (raffenet) [inline src/mpl/src/shm/mpl_shm_mmap.c]: initialization here really doesn't do anything since we set the value on the next line, so this change is unneeded. in fact, we could probably remove the initialization in a later commit if we wanted.

File: src/mpl/src/shm/mpl_shm_mmap.c [hunk 2]
```diff
@@ -149,7 +150,7 @@ int MPL_shm_seg_create_and_attach(MPL_shm_hnd_t hnd, intptr_t seg_sz,
  */
 int MPL_shm_seg_attach(MPL_shm_hnd_t hnd, intptr_t seg_sz, char **shm_addr_ptr, int offset)
 {
-    int rc = 0;
+    int rc = -1;
```
Comment (raffenet) [inline src/mpl/src/shm/mpl_shm_mmap.c]: ditto above.

File: src/mpl/src/shm/mpl_shm_mmap.c (hunk 1 above)
Comment (jain-surabhi-23) [reply src/mpl/src/shm/mpl_shm_mmap.c]: Ok. I will amend the commit and do `return MPL_shm_seg_create_attach_templ(...)` instead.

**Comment by jain-surabhi-23** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/all

**Comment by jain-surabhi-23** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/all

**Comment by jain-surabhi-23** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/all

File: src/mpl/src/shm/mpl_shm.c [hunk 3]
```diff
@@ -18,9 +18,7 @@
  */
 int MPL_shm_hnd_serialize(char *str, MPL_shm_hnd_t hnd, int str_len)
 {
-    int rc = -1;
-    rc = MPLI_shm_ghnd_get_by_val(hnd, str, str_len);
-    return rc;
+    return MPLI_shm_ghnd_get_by_val(hnd, str, str_len);
 }
```
Comment (minsii) [inline src/mpl/src/shm/mpl_shm.c]: Should this change go to commit: `mpl/shm: Remove unnecessary variables for return` ?

File: src/mpl/src/shm/mpl_shm.c [hunk 4]
```diff
@@ -53,9 +51,8 @@ int MPL_shm_hnd_deserialize(MPL_shm_hnd_t hnd, const char *str_hnd, size_t str_h
 
 int MPL_shm_hnd_get_serialized_by_ref(MPL_shm_hnd_t hnd, char **str_ptr)
 {
-    int mpi_errno = 0;
     *str_ptr = (char *) MPLI_shm_ghnd_get_by_ref(hnd);
-    return mpi_errno;
+    return 0;
 }
```
Comment (minsii) [inline src/mpl/src/shm/mpl_shm.c]: Should this change go to commit: mpl/shm: Remove unnecessary variables for return ?

File: src/mpl/src/shm/mpl_shm.c [hunk 5]
```diff
@@ -107,5 +98,5 @@ int MPL_shm_hnd_finalize(MPL_shm_hnd_t * hnd_ptr)
 
     *hnd_ptr = MPL_SHM_HND_INVALID;
 
-    return mpi_errno;
+    return 0;
 }
```
Comment (minsii) [inline src/mpl/src/shm/mpl_shm.c]: Same here.

**Comment by jain-surabhi-23** [comment]
@minsii I made the requested changes.

File: src/mpl/src/shm/mpl_shm_mmap.c [hunk 6]
```diff
@@ -47,7 +47,8 @@ static inline int MPL_shm_seg_create_attach_templ(MPL_shm_hnd_t hnd, intptr_t se
                                                   void **shm_addr_ptr, int offset, int flag,
                                                   MPL_memory_class class)
 {
-    MPLI_shm_lhnd_t lhnd = -1, rc = -1;
+    MPLI_shm_lhnd_t lhnd = -1;
+    int rc = -1, failure = 0;
```
Comment (raffenet) [inline src/mpl/src/shm/mpl_shm_mmap.c]: Rather than use a `failure` variable, can we add an `fn_fail` label and set `rc` to -1 there? That is more closely aligned with other parts of MPICH/MPL.

Comment (jain-surabhi-23) [reply src/mpl/src/shm/mpl_shm_mmap.c]: We need a second variable `failure` because we want to close the handle even in the case of an error, but return `-1` and not the output of `rc = MPLI_shm_lhnd_close(hnd);`.

Comment (raffenet) [reply src/mpl/src/shm/mpl_shm_mmap.c]: Ah, I see. Well this is still an improvement so I suppose we shouldn't block on it. @minsii do you have any comments? The other patches look fine to me.

Comment (minsii) [reply src/mpl/src/shm/mpl_shm_mmap.c]: Actually I prefer the code style suggested by Ken as it is consistent with other MPICH code. Instead of the `failure` branch, we could do:
```
int rc = 0, rc_close = 0; /* init */
...
if (error returned)
    goto fn_fail;

fn_exit:
    if (MPLI_shm_lhnd_is_valid(hnd)) {
        rc_close = MPLI_shm_lhnd_close(hnd);
    }
    return (rc == -1) ? rc : rc_close;
fn_fail:
    rc = -1;
    goto fn_exit;
```

File: src/mpid/ch4/src/ch4_impl.h [hunk 7]
```diff
@@ -962,28 +962,24 @@ static inline int MPIDI_CH4U_allocate_shm_segment(MPIR_Comm * shm_comm_ptr,
                                                   void **base_ptr)
 {
     MPIR_Errflag_t errflag = MPIR_ERR_NONE;
-    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno = MPI_SUCCESS, mpl_err = -1;
```
Comment (minsii) [inline src/mpid/ch4/src/ch4_impl.h]: Can we initialize `mpl_err = 0` to be consistent with `mpi_errno ` initialization ?

File: src/mpid/ch4/src/ch4_impl.h [hunk 8]
```diff
@@ -1036,18 +1029,16 @@ static inline int MPIDI_CH4U_destroy_shm_segment(MPI_Aint shm_segment_len,
                                                  MPL_shm_hnd_t * shm_segment_hdl_ptr,
                                                  void **base_ptr)
 {
-    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno = MPI_SUCCESS, mpl_err = -1;
```
Comment (minsii) [inline src/mpid/ch4/src/ch4_impl.h]: Same comment for `mpl_err ` initialization here.

File: src/mpid/common/shm/mpidu_shm_alloc.c [hunk 9]
```diff
@@ -137,7 +137,7 @@ int MPIDU_shm_seg_commit(MPIDU_shm_seg_t * memory, MPIDU_shm_barrier_t ** barrie
                          int num_local, int local_rank, int local_procs_0, int rank,
                          MPL_memory_class class)
 {
-    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno = MPI_SUCCESS, mpl_err = -1;
     int pmi_errno;
```
Comment (minsii) [inline src/mpid/common/shm/mpidu_shm_alloc.c]: Same comment for `mpl_err` initialization

File: src/mpid/common/shm/mpidu_shm_alloc.c [hunk 10]
```diff
@@ -656,18 +638,17 @@ int MPIDU_shm_seg_commit(MPIDU_shm_seg_t * memory, MPIDU_shm_barrier_t ** barrie
 #define FCNAME MPL_QUOTE(FUNCNAME)
 int MPIDU_shm_seg_destroy(MPIDU_shm_seg_t * memory, int num_local)
 {
-    int mpi_errno = MPI_SUCCESS;
+    int mpi_errno = MPI_SUCCESS, mpl_err = -1;
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDU_SHM_SEG_DESTROY);
```
Comment (minsii) [inline src/mpid/common/shm/mpidu_shm_alloc.c]: Same `mpl_err ` initialization value comment.

File: src/mpl/src/shm/mpl_shm_mmap.c (hunk 6 above)
Comment (jain-surabhi-23) [reply src/mpl/src/shm/mpl_shm_mmap.c]: I made this change

**Comment by jain-surabhi-23** [comment]
@minsii The requested changes are made.

File: src/mpl/src/shm/mpl_shm_sysv.c [hunk 11]
```diff
@@ -44,12 +50,13 @@ static inline int MPL_shm_seg_create_attach_templ(MPL_shm_hnd_t hnd, intptr_t se
...
+        if (*shm_addr_ptr == (void *) -1) {
+            rc = -1;
+        }
     }
 
   fn_exit:
     return rc;
-  fn_fail:
-    goto fn_exit;
 }
```
Comment (minsii) [inline src/mpl/src/shm/mpl_shm_sysv.c]: Should this change be in a separate commit ? I.e., remove unused code ?

File: src/mpl/src/shm/mpl_shm_win.c [hunk 12]
```diff
@@ -31,30 +31,42 @@ static inline int MPL_shm_seg_create_attach_templ(MPL_shm_hnd_t hnd, intptr_t se
...
+        if (*shm_addr_ptr == NULL) {
+            rc = -1;
+        }
     }
 
   fn_exit:
     return rc;
-  fn_fail:
-    goto fn_exit;
 }
```
Comment (minsii) [inline src/mpl/src/shm/mpl_shm_win.c]: Should this change be in a separate commit ? I.e., remove unused code ?

Comment (jain-surabhi-23) [reply src/mpl/src/shm/mpl_shm_win.c]: Done

File: src/mpl/src/shm/mpl_shm_sysv.c (hunk 11 above)
Comment (jain-surabhi-23) [reply src/mpl/src/shm/mpl_shm_sysv.c]: Done

**Comment by minsii** [comment]
Thanks. Will merge it after a final test.

**Comment by minsii** [comment]
test:jenkins/ch4/most
test:jenkins/ch3/all
